release/src/router/iproute2-3.x/man/man8/lnstat.8
.TH LNSTAT 8
.SH NAME
lnstat \- unified linux network statistics
.SH SYNOPSIS
.B lnstat
.RI [ options ]
.SH DESCRIPTION
This manual page documents briefly the
.B lnstat
command.
.PP
\fBlnstat\fP is a generalized and more feature-complete replacement for the old rtstat program.
In addition to routing cache statistics, it supports any kind of statistics the linux kernel
exports via a file in /proc/net/stat/.
.SH OPTIONS
lnstat supports the following options.
.TP
.B \-h, \-\-help
Show summary of options.
.TP
.B \-V, \-\-version
Show version of program.
.TP
.B \-c, \-\-count <count>
Print <count> number of intervals.
.TP
.B \-d, \-\-dump
Dump list of available files/keys.
.TP
.B \-f, \-\-file <file>
Statistics file to use.
.TP
.B \-i, \-\-interval <intv>
Set interval to 'intv' seconds.
.TP
.B \-j, \-\-json
Display results in JSON format
.TP
.B \-k, \-\-keys k,k,k,...
Display only keys specified.
.TP
.B \-s, \-\-subject [0-2]
Specify display of subject/header. '0' means no header at all, '1' prints a header only at start of the program and '2' prints a header every 20 lines.
.TP
.B \-w, \-\-width n,n,n,...
Width for each field.
.SH USAGE EXAMPLES
.TP
.B # lnstat -d
Get a list of supported statistics files.
.TP
.B # lnstat -k arp_cache:entries,rt_cache:in_hit,arp_cache:destroys
Select the specified files and keys.
.TP
.B # lnstat -i 10
Use an interval of 10 seconds.
.TP
.B # lnstat -f ip_conntrack
Use only the specified file for statistics.
.TP
.B # lnstat -s 0
Do not print a header at all.
.TP
.B # lnstat -s 20
Print a header at start and every 20 lines.
.TP
.B # lnstat -c -1 -i 1 -f rt_cache -k entries,in_hit,in_slow_tot
Display statistics for keys entries, in_hit and in_slow_tot of field rt_cache every second.
.SH SEE ALSO
.BR ip (8),
and /usr/share/doc/iproute-doc/README.lnstat (package iproute-doc on Debian)
.br
.SH AUTHOR
lnstat was written by Harald Welte <laforge@gnumonks.org>.
.PP
This manual page was written by Michael Prokop <mika@grml.org> for the Debian project (but may be used by others).


Changelog.txt
Asuswrt-Merlin Changelog
========================

380.59 (xx-xxx-2016)
    - NEW: Merged with 380_2697 GPL
    - NEW: Option on OpenVPN client/server page to reset them back to the factory
           default settings.
    - CHANGED: WAN -> NAT Passthrough now allows you to determine wether or not to load
               the NAT helper module for h323, rtsp and sip.  Asus's old behaviour is
               "Enabled + NAT Helper".
    - CHANGED: DNSFilter client dropdown now uses Asus's new one integrated with networkmap.
    - CHANGED: minidlna now supports refreshing an existing database, so the Tweak setting
               was updated accordingly
    - CHANGED: Enable SPNEGO support in Samba
    - CHANGED: Integrated Asus's networkmap into the DHCP reservations page
    - CHANGED: Updated Tor to 0.2.7.6
    - FIXED: Daily/Monthly traffic monitoring shows invalid values on the RT-AC88U/3100/5300,
             even with CTF disabled.
    - FIXED: WPS wasn't working on the RT-AC3200


380.58 (20-Mar-2016)
    - NEW: Merged with 380_1354 GPL
    - NEW: Added Tweaks and Hacks settings to Tools -> Other Settings.
           These are UNSUPPORTED tweaks, intended mostly for 
           experimentation, or very specific situations.  If unsure how
           to apply these, manually reboot after changing them.
           One of new settings there lets you disable hourly network
           rescans, to resolve issues with NAS/printers coming out
           of sleep every hour.
    - NEW: Added setting to configure OpenVPN's auth digest algo.
    - NEW: Added setting to configure OpenVPN's logging verbosity.
           Note that this setting is global to all clients/servers.
    - CHANGED: Updated OpenVPN to 2.3.10
    - CHANGED: Updated openssl to 1.0.2g
    - CHANGED: Updated miniupnpd to 1.9.20160222
    - CHANGED: Updated udpxy to 1.0-build 23-10 (backport from GPL 
               380_2345)
    - CHANGED: if you set an OpenVPN client DNS mode to "Exclusive"
               and you enable policy-based routing, then those policies
               will also determine which DNS to use (the tunnel's or
               the ISP's).  This is based on DNSFilter's technology.
               You no longer need to use DNSFilter to control
               the DNS used by your OpenVPN clients.
    - CHANGED: Made OpenVPN traffic bypass CTF, which resolves
               some throughput issues with it
    - CHANGED: Disabled X11 Forwarding support in Dropbear,
               for security reasons.
    - FIXED: PPTP static route handling script was broken
    - FIXED: minidlna would check for the wrong database filename
             at start time
    - FIXED: Wrong status shown for VPN Client 3
    - FIXED: OpenVPN clients were run on the wrong CPU cores.
             Now, odd instances correctly run on the second core.
    - FIXED: Using DNSFilter with default mode set to "router" would
             prevent using the router for IPv6 lookups.
    - FIXED: Account limit wasn't properly allowing up to 10 
             clients for SMB/FTP (patch by vit9696)
    - FIXED: Having multiple OpenVPN clients configured with
             multiple "Accept DNS configuration" modes would
             only apply the last client's setting.  Now, we
             apply the most restrictive setting of all
             configured clients.
    - FIXED: RT-AC68U 2.4 GHz was broken if CTF was disabled
             (downgraded wifi driver to  6.37.14.105)
    - FIXED: Diasbling the SIP NAT helper would also drop all port 5060 
             traffic.  Some users need to keep the SIP helper disabled 
             with their SIP client.  Reverted that GPL 858 change.


380.57 (24-Dec-2015)
    - NEW: Merged with 380_1031 GPL
    - NEW: Added RT-AC3100 and RT-AC5300 support
    - NEW: Added RT-AC68U HW Revision C1 support
    - NEW: Backup/Restore of the content of the JFFS
           partition (under Administration Restore/Save Settings)
    - NEW: Added DNSSEC support.  Can be enabled under LAN -> DHCP.
    - NEW: Added custom/postconf support for igmpproxy.conf.
    - CHANGED: Increased user account limit from 16 to 32 on
               the VPN server pages.
    - CHANGED: Updated e2fsprogs to 1.42.13
    - CHANGED: Increased maximum entries in Parental Control
               (time scheduler) to 32.
    - CHANGED: Updated miniupnpd to 1.9.20151119.
    - CHANGED: Updated Openssl to 1.0.2e.
    - CHANGED: Downgraded Dropbear to 2014.66, too many issues in
               the newer releases.
    - CHANGED: Improvements to VPN Status page
    - FIXED: CTF not automatically disabled when enabling IPTraffic.
    - FIXED: Openvpn clients 3 through 5 were all run on the first
             CPU core.  They are now properly alternated like the
             first two (odd on CPU1, even on CPU0)
    - FIXED: smb.log generated by networkmap could fill up RAM
    - FIXED: upnpc_xml.log generated by miniupnpc could fill up RAM
    - FIXED: Inconsistant names used on IPTraffic and Sysinfo page.
             Now, we give priority to any description manually entered
             on the networkmap, followed by static hostname, then any
             current (lease) hostname.
   - FIXED: MAC queries sent to the OUI database were broken due to
            changes on the IEEE website
   - FIXED: Applying changes to OpenVPN client page would start the
            client even if it was disabled/stopped.


378.56_2 (2-Nov-2015)
    - CHANGED: Reverted the memory buffering optimization
               for ARM devices, as people keep panicking
               over the lower amount of free RAM.  You can 
               manually re-enable the optimization by setting
               "drop_caches=0" in nvram.
    - CHANGED: Allow using a port < 1024 for http(s) webui
               interface.
    - FIXED: EMF wasn't working on AC56/AC68/AC87.

release/src/router/iproute2-3.x/misc/lnstat.c
/* lnstat - Unified linux network statistics
 *
 * Copyright (C) 2004 by Harald Welte <laforge@gnumonks.org>
 *
 * Development of this code was funded by Astaro AG, http://www.astaro.com/
 *
 * Based on original concept and ideas from predecessor rtstat.c:
 *
 * Copyright 2001 by Robert Olsson <robert.olsson@its.uu.se>
 *                                 Uppsala University, Sweden
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 */

/* Maximum number of fields that can be displayed */
#define MAX_FIELDS		128

/* Maximum number of header lines */
#define HDR_LINES		10

/* default field width if none specified */
#define FIELD_WIDTH_DEFAULT	8
#define FIELD_WIDTH_MAX		20

#define DEFAULT_INTERVAL	2

#define HDR_LINE_LENGTH		(MAX_FIELDS*FIELD_WIDTH_MAX)

#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <getopt.h>

#include "lnstat.h"

static struct option opts[] = {
	{ "version", 0, NULL, 'V' },
	{ "count", 1, NULL, 'c' },
	{ "dump", 0, NULL, 'd' },
	{ "json", 0, NULL, 'j' },
	{ "file", 1, NULL, 'f' },
	{ "help", 0, NULL, 'h' },
	{ "interval", 1, NULL, 'i' },
	{ "keys", 1, NULL, 'k' },
	{ "subject", 1, NULL, 's' },
	{ "width", 1, NULL, 'w' },
};

static int usage(char *name, int exit_code)
{
	fprintf(stderr, "%s Version %s\n", name, LNSTAT_VERSION);
	fprintf(stderr, "Copyright (C) 2004 by Harald Welte "
			"<laforge@gnumonks.org>\n");
	fprintf(stderr, "This program is free software licensed under GNU GPLv2"
			"\nwith ABSOLUTELY NO WARRANTY.\n\n");
	fprintf(stderr, "Parameters:\n");
	fprintf(stderr, "\t-V --version\t\tPrint Version of Program\n");
	fprintf(stderr, "\t-c --count <count>\t"
			"Print <count> number of intervals\n");
	fprintf(stderr, "\t-d --dump\t\t"
			"Dump list of available files/keys\n");
	fprintf(stderr, "\t-j --json\t\t"
			"Display in JSON format\n");
	fprintf(stderr, "\t-f --file <file>\tStatistics file to use\n");
	fprintf(stderr, "\t-h --help\t\tThis help message\n");
	fprintf(stderr, "\t-i --interval <intv>\t"
			"Set interval to 'intv' seconds\n");
	fprintf(stderr, "\t-k --keys k,k,k,...\tDisplay only keys specified\n");
	fprintf(stderr, "\t-s --subject [0-2]\t?\n");
	fprintf(stderr, "\t-w --width n,n,n,...\tWidth for each field\n");
	fprintf(stderr, "\n");

	exit(exit_code);
}

struct field_param {
	const char *name;
	struct lnstat_field *lf;
	struct {
		unsigned int width;
	} print;
};

struct field_params {
	unsigned int num;
	struct field_param params[MAX_FIELDS];
};

static void print_line(FILE *of, const struct lnstat_file *lnstat_files,
		       const struct field_params *fp)
{
	int i;

	for (i = 0; i < fp->num; i++) {
		const struct lnstat_field *lf = fp->params[i].lf;

		fprintf(of, "%*lu|", fp->params[i].print.width, lf->result);
	}
	fputc('\n', of);
}

static void print_json(FILE *of, const struct lnstat_file *lnstat_files,
		       const struct field_params *fp)
{
	int i;
	const char *sep;
	const char *base = NULL;

	fputs("{\n", of);
	for (i = 0; i < fp->num; i++) {
		const struct lnstat_field *lf = fp->params[i].lf;

		if (!base || lf->file->basename != base) {
			if (base) fputs("},\n", of);
			base = lf->file->basename;
			sep = "\n\t";
			fprintf(of, "    \"%s\":{", base);
		}
		fprintf(of, "%s\"%s\":%lu", sep,
			lf->name, lf->result);
		sep = ",\n\t";
	}
	fputs("}\n}\n", of);
}

/* find lnstat_field according to user specification */
static int map_field_params(struct lnstat_file *lnstat_files,
			    struct field_params *fps, int interval)
{
	int i, j = 0;
	struct lnstat_file *lf;

	/* no field specification on commandline, need to build default */
	if (!fps->num) {
		for (lf = lnstat_files; lf; lf = lf->next) {
			for (i = 0; i < lf->num_fields; i++) {
				fps->params[j].lf = &lf->fields[i];
				fps->params[j].lf->file->interval.tv_sec =
								interval;
				if (!fps->params[j].print.width)
					fps->params[j].print.width =
							FIELD_WIDTH_DEFAULT;

				if (++j >= MAX_FIELDS - 1) {
					fprintf(stderr,
						"WARN: MAX_FIELDS (%d) reached,"
						" truncating number of keys\n",
						MAX_FIELDS);
					goto full;
				}
			}
		}
	full:
		fps->num = j;
		return 1;
	}

	for (i = 0; i < fps->num; i++) {
		fps->params[i].lf = lnstat_find_field(lnstat_files,
						      fps->params[i].name);
		if (!fps->params[i].lf) {
			fprintf(stderr, "Field `%s' unknown\n",
				fps->params[i].name);
			return 0;
		}
		fps->params[i].lf->file->interval.tv_sec = interval;
		if (!fps->params[i].print.width)
			fps->params[i].print.width = FIELD_WIDTH_DEFAULT;
	}
	return 1;
}

struct table_hdr {
	int num_lines;
	char *hdr[HDR_LINES];
};

static struct table_hdr *build_hdr_string(struct lnstat_file *lnstat_files,
					  struct field_params *fps,
					  int linewidth)
{
	int h,i;
	static struct table_hdr th;
	int ofs = 0;

	for (i = 0; i < HDR_LINES; i++) {
		th.hdr[i] = malloc(HDR_LINE_LENGTH);
		memset(th.hdr[i], 0, HDR_LINE_LENGTH);
	}

	for (i = 0; i < fps->num; i++) {
		char *cname, *fname = fps->params[i].lf->name;
		unsigned int width = fps->params[i].print.width;

		snprintf(th.hdr[0]+ofs, width+2, "%*.*s|", width, width,
			 fps->params[i].lf->file->basename);

		cname = fname;
		for (h = 1; h < HDR_LINES; h++) {
			if (cname - fname >= strlen(fname))
				snprintf(th.hdr[h]+ofs, width+2,
					 "%*.*s|", width, width, "");
			else {
				th.num_lines = h+1;
				snprintf(th.hdr[h]+ofs, width+2,
					 "%*.*s|", width, width, cname);
			}
			cname += width;
		}
		ofs += width+1;
	}
	/* fill in spaces */
	for (h = 1; h <= th.num_lines; h++) {
		for (i = 0; i < ofs; i++) {
			if (th.hdr[h][i] == '\0')
				th.hdr[h][i] = ' ';
		}
	}

	return &th;
}

static int print_hdr(FILE *of, struct table_hdr *th)
{
	int i;

	for (i = 0; i < th->num_lines; i++) {
		fputs(th->hdr[i], of);
		fputc('\n', of);
	}
	return 0;
}


int main(int argc, char **argv)
{
	struct lnstat_file *lnstat_files;
	const char *basename;
	int i, c;
	int interval = DEFAULT_INTERVAL;
	int hdr = 2;
	enum {
		MODE_DUMP,
		MODE_JSON,
		MODE_NORMAL,
	} mode = MODE_NORMAL;
	unsigned long count = 1;
	struct table_hdr *header;
	static struct field_params fp;
	int num_req_files = 0;
	char *req_files[LNSTAT_MAX_FILES];

	/* backwards compatibility mode for old tools */
	basename = strrchr(argv[0], '/');
	if (basename)
		basename += 1;	  /* name after slash */
	else
		basename = argv[0]; /* no slash */

	if (!strcmp(basename, "rtstat")) {
		/* rtstat compatibility mode */
		req_files[0] = "rt_cache";
		num_req_files = 1;
	} else if (!strcmp(basename, "ctstat")) {
		/* ctstat compatibility mode */
		req_files[0] = "ip_conntrack";
		num_req_files = 1;
	}

	while ((c = getopt_long(argc, argv,"Vc:djf:h?i:k:s:w:",
				opts, NULL)) != -1) {
		int len = 0;
		char *tmp, *tok;

		switch (c) {
		case 'c':
			count = strtoul(optarg, NULL, 0);
			break;
		case 'd':
			mode = MODE_DUMP;
			break;
		case 'j':
			mode = MODE_JSON;
			break;
		case 'f':
			req_files[num_req_files++] = strdup(optarg);
			break;
		case '?':
		case 'h':
			usage(argv[0], 0);
			break;
		case 'i':
			sscanf(optarg, "%u", &interval);
			break;
		case 'k':
			tmp = strdup(optarg);
			if (!tmp)
				break;
			for (tok = strtok(tmp, ",");
			     tok;
			     tok = strtok(NULL, ",")) {
				if (fp.num >= MAX_FIELDS) {
					fprintf(stderr,
						"WARN: too many keys"
						" requested: (%d max)\n",
						MAX_FIELDS);
					break;
				}
				fp.params[fp.num++].name = tok;
			}
			break;
		case 's':
			sscanf(optarg, "%u", &hdr);
			break;
		case 'w':
			tmp = strdup(optarg);
			if (!tmp)
				break;
			i = 0;
			for (tok = strtok(tmp, ",");
			     tok;
			     tok = strtok(NULL, ",")) {
				len  = strtoul(tok, NULL, 0);
				if (len > FIELD_WIDTH_MAX)
					len = FIELD_WIDTH_MAX;
				fp.params[i].print.width = len;
				i++;
			}
			if (i == 1) {
				for (i = 0; i < MAX_FIELDS; i++)
					fp.params[i].print.width = len;
			}
			break;
		default:
			usage(argv[0], 1);
			break;
		}
	}

	lnstat_files = lnstat_scan_dir(PROC_NET_STAT, num_req_files,
				       (const char **) req_files);

	switch (mode) {
	case MODE_DUMP:
		lnstat_dump(stderr, lnstat_files);
		break;

	case MODE_NORMAL:
	case MODE_JSON:
		if (!map_field_params(lnstat_files, &fp, interval))
			exit(1);

		header = build_hdr_string(lnstat_files, &fp, 80);
		if (!header)
			exit(1);

		if (interval < 1 )
			interval = 1;

		for (i = 0; i < count; i++) {
			lnstat_update(lnstat_files);
			if (mode == MODE_JSON)
				print_json(stdout, lnstat_files, &fp);
			else {
				if  ((hdr > 1 &&
				      (! (i % 20))) || (hdr == 1 && i == 0))
					print_hdr(stdout, header);
				print_line(stdout, lnstat_files, &fp);
			}
			fflush(stdout);
			if (i < count - 1)
				sleep(interval);
		}
		break;
	}

	return 1;
}

